src/app/content/utils/seoUtils.ts

```diff
@@ -9,20 +12,166 @@ import {
+  bookTitle: string;
+  parentPrefix: string;
+}
+
+const domParser = new DOMParser();
+
+export const getParentPrefix = (node: LinkedArchiveTreeNode | undefined, includeTitle: boolean = false): string => {
   if (!node) {
     return '';
   }
 
   if (archiveTreeSectionIsChapter(node)) {
-    const number = getArchiveTreeSectionNumber(node);
-    return `Ch. ${number} `;
+    const number = getArchiveTreeSectionNumber(node).trim();
+    const name = getArchiveTreeSectionTitle(node);
+    return includeTitle ? `Ch. ${number}. ${name}` : `Ch. ${number} `;
   }
 
   return archiveTreeSectionIsBook(node.parent)
     ? getArchiveTreeSectionTitle(node) + ' '
-    : getParentPrefix(node.parent);
+    : getParentPrefix(node.parent, includeTitle);
+
+};
+
+const hideMath = (node: Element) => {
+  if (!node) {
+    return '';
+  }
+
+  node.querySelectorAll('math, .os-math-in-para').forEach((el: Element) => {
+    el.outerHTML = '...';
+  });
+  return node.textContent;
+};
+
+export const getTextContent = (str: string) => {
+  const parsed = domParser.parseFromString(str, 'text/html');
+  const text = parsed.body.textContent;
+  return text;
+};
+
+const removeExcludedContent = (node: HTMLElement) => {
+  if (!node) {
+    return null;
+  }
+  const excludedContent = node.querySelectorAll(
+    '[data-type="abstract"], .learning-objectives, .chapter-objectives, .be-prepared, .os-teacher'
+    ) || [];
+
+  for (let i = 0; i <= excludedContent.length; i++) {
+    if (excludedContent[i]) {
+      excludedContent[i].remove();
+    }
+  }
+};
+
+export const generateExcerpt = (str: string) => {
+  return str.replace(/\n+/g, ' ').replace(/\s+/g, ' ').trim().substring(0, 152) + '...';
+};
+
+const getPageType = (node: HTMLElement, values: DescriptionTemplateValues): PageTypes => {
+  const {parentType, parentPrefix, pageTitle} = values;
+  const nodeClasses = node.classList;
+  const nodeType = node.getAttribute('data-type');
+
+  if (nodeType === 'page' && !nodeClasses.contains('appendix')) {
+    return 'page';
+  } else if (pageTitle === parentPrefix) {
+    return 'eob-page';
+  } else if (
+    nodeClasses.contains('os-solution-container')
+      || nodeClasses.contains('os-solutions-container')
+    ) {
+    return 'answer-key';
+  } else if (parentType !== 'chapter' && parentType !== 'book') {
+    return 'subpage';
+  } else if (pageTitle !== parentPrefix) {
+    return 'eoc-page';
+  } else {
+    throw new Error('no page type detected');
+  }
+};
+
+const getPageDescriptionFromContent = (node: HTMLElement): string | null => {
+  if (!node) {
+    return null;
+  }
+  removeExcludedContent(node);
+  const paragraphs = [...Array.from(node.querySelectorAll('section>p')), ...Array.from(node.querySelectorAll('p'))];
+
+  if (!paragraphs) {
+    return null;
+  }
+
+  const foundByLength = Array.from(paragraphs).find((p) => p.textContent ? p.textContent.length >= 90 : null);
+  const mathless = foundByLength ? hideMath(foundByLength) : null;
```

Comment:
maybe do hidemath before the length trimming, i think that'll give better results